### ISSUE TITLE
TORCHDYNAMO_DYNAMIC_SHAPES=1 expects sizes returned from graph to be tuples, but AOTAutograd doesn't agree

### ISSUE THREAD

**Issue by ezyang** [pytorch/torchdynamo#1466]
Repro: `TORCH_SHOW_CPP_STACKTRACES=1 TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 TORCH_MONKEYPATCH_GETITEM=1 AOT_FX_GRAPHS=1 python benchmarks/torchbench.py --only hf_T5 --accuracy --backend aot_eager --training)` with symbolic-shapes branch https://github.com/pytorch/pytorch/pull/84246

We end up with this AOTAutograd graph:

```
====== Forward graph ======                                                                                                                                
class GraphModule(torch.nn.Module):                                                                                                                        
    def forward(self, primals_1, primals_2, primals_3):                                                                                                    
                                                                                                                                                           
        # Module stack: {}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py:925, code: input_ids
 = input_ids.view(-1, input_shape[-1])                                                                                                                     
        sym_size = torch.ops.aten.sym_size(primals_2, 1)                                                                                                   
        view = torch.ops.aten.view.default(primals_2, [-1, sym_size])                                                                                      
                                                                             
        # Module stack: {'self_embed_tokens': 'Embedding'}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/mod
eling_t5.py:934, code: inputs_embeds = self.embed_tokens(input_ids)          
        embedding = torch.ops.aten.embedding.default(primals_1, view);  primals_1 = None                                                                   
                                                                             
        # Module stack: {}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py:945, code: attention
_mask = torch.ones(batch_size, mask_seq_length).to(inputs_embeds.device)                                                                                   
        sym_size_1 = torch.ops.aten.sym_size(primals_2, 0);  primals_2 = None                                                                              
        ones = torch.ops.aten.ones.default([sym_size_1, sym_size], device = device(type='cpu'), pin_memory = False)                              
        detach = torch.ops.aten.detach.default(ones);  ones = None                                                                                         
        detach_1 = torch.ops.aten.detach.default(detach);  detach = None                                                                                   
        _to_copy = torch.ops.aten._to_copy.default(detach_1, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0));  detach
_1 = None                                                                    
                                                                                                                                                           
        # Module stack: {}, File: /home/ezyang/local/pytorch-tmp-env/lib/python3.9/site-packages/transformers/models/t5/modeling_t5.py:948, code: encoder_a
ttention_mask = torch.ones(                                                                                                                                
        sym_size_2 = torch.ops.aten.sym_size(primals_3, 1);  primals_3 = None                                                                              
        ones_1 = torch.ops.aten.ones.default([sym_size_1, sym_size_2], dtype = torch.int64, device = device(type='cuda', index=0), pin_memory = False);  sy
m_size_2 = None                                                              
        detach_2 = torch.ops.aten.detach.default(ones_1);  ones_1 = None                                                                                   
        detach_3 = torch.ops.aten.detach.default(detach_2);  detach_2 = None                                                                               
        return [_to_copy, sym_size_1, sym_size, detach_3, embedding, view] 

aot output: (tensor([[1., 1., 1.,  ..., 1., 1., 1.],                                                                                                       
        [1., 1., 1.,  ..., 1., 1., 1.]], device='cuda:0',                                                                                                  
       grad_fn=<CompiledFunctionBackward>), 2, 1024, tensor([[1, 1, 1,  ..., 1, 1, 1],                                                                     
        [1, 1, 1,  ..., 1, 1, 1]], device='cuda:0'))  
```

but Dynamo is expecting (2, 1024)

Comments from @jansel :

Actually, perhaps we should just change `DynamicShapeVariable(len=2)` to `TupleVariable([DynamicShapeVariable, DynamicShapeVariable])`
There is actually a SizeVariable() in [lists.py](https://lists.py/). This is only a dynamic_shapes=True thing.
(dynamic_shapes=True is less tested)

We live debugged this on the stream at https://www.twitch.tv/edwardzyang

**Comment by jansel** [comment]
This fix should involve changing these lines:
https://github.com/pytorch/torchdynamo/blob/7350f88e00a4e03c6eb591546727c25151e5ca56/torchdynamo/variables/tensor.py#L180-L187

To return a `torchdynamo.variables.SizeVariable([DynamicShapeVariable(), ...])` (which is a type of TupleVariable).

**Comment by jansel** [comment]
Actually it would look somewhat similar to the static shapes case right below:
https://github.com/pytorch/torchdynamo/blob/7350f88e00a4e03c6eb591546727c25151e5ca56/torchdynamo/variables/tensor.py#L196-L200

**Comment by ezyang** [comment]
```
diff --git a/torchdynamo/variables/tensor.py b/torchdynamo/variables/tensor.py
index 0ec864de..6d562348 100644
--- a/torchdynamo/variables/tensor.py
+++ b/torchdynamo/variables/tensor.py
@@ -178,13 +178,14 @@ class TensorVariable(VariableTracker):
             options.update(specialized_props)
             return cls(proxy, **options)
         elif (
-            istype(example_value, (torch.Size, int, bool, float))
+            istype(example_value, (int, bool, float))
             and config.dynamic_shapes
         ):
             proxy.node.meta["example_value"] = example_value
-            if isinstance(example_value, torch.Size):
-                options["dyn_shape_len"] = len(example_value)
             return DynamicShapeVariable(proxy, type(example_value), **options)
+        elif istype(example_value, torch.Size) and config.dynamic_shapes:
+            sizes = [DynamicShapeVariable(proxy.size(i), int) for i in range(len(example_value))]
+            return SizeVariable(sizes, **options)
         elif istype(example_value, int) and proxy.node.target in (
             torch.seed,
             operator.mod,
```

just need to write a test
